tests/unit/src/Internal/Service/PaymentProcessingServiceTest.php

```diff
@@ -126,8 +143,40 @@ public function test_process_payment_happy_path_without_mock_builder() {
 		$this->assertSame( $mock_completed_state, $result );
 	}
 
+	/**
+	 * Test the process payment when mode not initialized.
+	 */
+	public function test_process_payment_mode_throws_exception() {
+		$this->mock_mode
+			->expects( $this->once() )
+			->method( 'is_test' )
+			->willThrowException( new Exception( 'Could not initialize' ) );
+
+		$sut = new PaymentProcessingService( $this->mock_state_factory, $this->mock_legacy_proxy, $this->mock_context_logger, $this->mock_logger, $this->mock_mode );
+
+		$mock_initial_state   = $this->createMock( InitialState::class );
+		$mock_completed_state = $this->createMock( CompletedState::class );
+
+		// Prepare a payment method for the context.
+		$_POST['payment_method']       = WC_Payment_Gateway_WCPay::GATEWAY_ID;
+		$_POST['wcpay-payment-method'] = 'pi_XYZ';
+
+		$this->mock_state_factory->expects( $this->once() )
+			->method( 'create_state' )
+			->with( InitialState::class, $this->isInstanceOf( PaymentContext::class ) )
+			->willReturn( $mock_initial_state );
+
+		$mock_initial_state->expects( $this->once() )
+			->method( 'start_processing' )
+			->with( $this->isInstanceOf( PaymentRequest::class ) )
+			->willReturn( $mock_completed_state );
+
+		// process completes as mode is used only for logging and meta.
+		$this->assertSame( $mock_completed_state, $sut->process_payment( 1 ) );
```

Comment:
Is there a reason to test all of `process_payment` just to see whether `create_state` handles the mode correctly? IMO as long as the exception is thrown, and `create_payment_context` handles it, that should be good enough.

Reply:
Any suggestion for how to examine that the un-mocked protected method `create_payment_context` handles the exception without errors?

Reply:
It's relatively easy to test `protected` or `private` methods. You can:

1. Create a temporary instance of the service, like how it's done on line 99 (within the `test_process_payment_happy_path` test).
2. Use `PHPUnit_Utils::call_method`, see https://github.com/Automattic/woocommerce-payments/blob/942dfb0579f0a9d27af1f409ff179ea403339031/tests/unit/src/Internal/Payment/State/InitialStateTest.php#L520 for reference.

Reply:
Changed